django/db/models/fields/related_descriptors.py

```diff
@@ -225,21 +225,38 @@ def __get__(self, instance, cls=None):
         try:
             rel_obj = self.field.get_cached_value(instance)
         except KeyError:
+            rel_obj = None
             has_value = None not in self.field.get_local_related_value(instance)
-            ancestor_link = (
-                instance._meta.get_ancestor_link(self.field.model)
-                if has_value
-                else None
-            )
-            if ancestor_link and ancestor_link.is_cached(instance):
+            current_instance = instance
+            keep_going = has_value
+            while keep_going:
+                ancestor_link = current_instance._meta.get_ancestor_link(
+                    self.field.model
+                )
                 # An ancestor link will exist if this field is defined on a
                 # multi-table inheritance parent of the instance's class.
-                ancestor = ancestor_link.get_cached_value(instance)
-                # The value might be cached on an ancestor if the instance
-                # originated from walking down the inheritance chain.
-                rel_obj = self.field.get_cached_value(ancestor, default=None)
-            else:
-                rel_obj = None
+                if (
+                    ancestor_link
+                    and (
+                        ancestor := ancestor_link.get_cached_value(
+                            current_instance, None
+                        )
+                    )
+                    is not None
+                ):
+                    # The value might be cached on an ancestor if the instance
+                    # originated from walking down the inheritance chain.
+                    rel_obj = self.field.get_cached_value(ancestor, default=None)
+                    if rel_obj is not None:
+                        # We found it in the cache
+                        keep_going = False
+                    else:
+                        # Keep going with further ancestors
+                        current_instance = ancestor
+                else:
+                    # no more ancestor or the ancestor instance is not cached
+                    rel_obj = None
+                    keep_going = False
```

Comment:
Looking at the `__set__` logic I wonder if we also need some special handling here in cases where it fiddles with `self.field.get_cached_value` 🤔

Reply:
Currently, setting the foreign key doesn't overwrite the parent's value. So this fails:
```python
def test_setting_fk_in_child_clears_parent_cache(self):
    obj = (
        Author.objects.select_related("authorwithage")
        .prefetch_related("first_book")
        .get(pk=self.child.pk)
    )
    obj.authorwithage.first_book = self.book2
    self.assertEqual(obj.first_book, self.book2)
```

But this already fails without my patch. Do you think it should be changed? I'm not sure if that would be classified as a breaking change.

Reply:
I think it can be kept as is for now if it doesn't change the behaviour. I wanted to make sure the change we made wasn't causing some subtle breakage in this regard but since it doesn't change the location of the cache (just how it's retrieved) we should be fine.